toolchain/lexer/tokenized_buffer_benchmark.cpp

```diff
@@ -0,0 +1,136 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <benchmark/benchmark.h>
+
+#include <forward_list>
+
+#include "absl/random/random.h"
+#include "common/check.h"
+#include "llvm/ADT/Sequence.h"
+#include "llvm/ADT/StringExtras.h"
+#include "toolchain/diagnostics/null_diagnostics.h"
+#include "toolchain/lexer/token_kind.h"
+#include "toolchain/lexer/tokenized_buffer.h"
+
+namespace Carbon::Testing {
+namespace {
+
+class LexerBenchHelper {
+ public:
+  explicit LexerBenchHelper(llvm::StringRef text)
+      : source_(MakeSourceBuffer(text)) {}
+
+  auto Lex() -> TokenizedBuffer {
+    DiagnosticConsumer& consumer = NullDiagnosticConsumer();
+    return TokenizedBuffer::Lex(source_, consumer);
+  }
+
+  auto DiagnoseErrors() -> llvm::StringRef {
+    auto buffer = TokenizedBuffer::Lex(source_, ConsoleDiagnosticConsumer());
+    CARBON_CHECK(buffer.has_errors())
+        << "Asked to diagnose errors but none found!";
+    // Return an empty string so this can be streamed into a check failure.
+    return "";
+  }
```

Comment:
I was looking at this, and had a related suggestion... The diagnostic consumer should only get called when has_errors is true, maybe you can pass ConsoleDiagnosticConsumer instead of NullDiagnosticConsumer and only check has_errors?